### PR TITLE
Verify that main info response returns correct product headers

### PR DESCRIPTION
**Pull request by swallez**
Follow-up to #73434

Ensures that High Level Rest Client is running against a verified
Elasticsearch. When the first request is sent on HLRC, a request to the
info endpoint is made first to verify the product identification and
version.

### CODE REVIEW

**Comment by swallez** [comment]
Build failures are because https://github.com/elastic/elasticsearch/pull/73434 has not been backported to `7.x` (see https://github.com/elastic/elasticsearch/pull/73434#issuecomment-856896383). The code is ready for review though.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:2017]: This took me a considerable amount of time to figure out what was actually happening in this method. Could you add some comments to help future readers determine what this method is doing?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:2043]: Why do we cancel the call when the future is complete? Won't the future be completed successfully but the original call still be cancelled?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:2113]: I believe the plan was if the request to `/` failed with a 401 or 403, then subsequent checks must be ignored. We check for 401 and 403 in the `onSuccess` handler, but aren't those considered exceptional cases which invoke the `onFailure` response handler?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java
Comment (dakrone) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:1033]: I think we need a check for when the root request is exceptional, throwing either a 401 or 403, and also if it throws something other than those status requests.

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks for working on this @swallez, I left some comments

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
Comment (swallez) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:2035]: The `cancelFlag` future is completed (and not cancelled) when the `Cancellable` returned as a result is cancelled. This is because `CompletableFuture` allows to listen to normal and exceptional completion but not on cancellation. Also there's `isCancelled` and `isCompletedExceptionally` but no `isCompletedSuccessfully`. So `isDone` works here, since the only thing we ever do to this future is to `complete(null)`.

The comments added in e0e0638 provide more details on this flag.

Comment (swallez) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:2035]: Actually the above comment is wrong: cancelling a `CompletableFuture` completes it exceptionnally with a `CancellationException`. I'll fix the comment, but woud like to keep the current behavior as we're not cancelling the flag but raising it.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
Comment (swallez) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:2048]: Not sure I understnad the question. The `cancelFlag` future is completed successfully when the method's result is cancelled. If/when that happens, we also need to cancel the asynchronous API call request.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
Comment (swallez) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:2131]: Good catch, LLRC will indeed raise an exception for non-nominal response status codes. I added them to the `"ignore"` list in d981a15 along with some tests.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java
Comment (swallez) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:1034]: See d981a15 for 401 & 403. Other failures of the info request are propagated as failures for the original API request. If `/` returns something other than 200, 401 or 403 then we have a bigger problem and we won't let API requests through, both because verification could not happen and also because they will most likely fail too.

**Review by dakrone — APPROVED** [review]
LGTM, thanks for iterating on this. I left a few other minor nits but nothing major.

**Review by jbaiera — CHANGES_REQUESTED** [review]
A quick note about compatibility coverage that should be addressed before merge:

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
Comment (jbaiera) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:2177]: I think we want to do the tagline validation here as well.

**Comment by swallez** [comment]
@dakrone I added some comments and refactored a bit in 115e9d1 and 94d88f7: `cancelFlag` is now a `cancellationForwarder` that is effectively cancelled.

@jbaiera I added build flavor and tagline validation in ceb527a and b87302d

Can you both please cross-check? Thanks!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:2192]: This needs to take into account `Build.Flavor.OSS`, because a 7.10 OSS build should still allow the HLRC to be used I think?

**Review by dakrone — CHANGES_REQUESTED** [review]
I think we need to support OSS builds here also, so we need to change the build flavor check

**Review by jbaiera — APPROVED** [review]
LGTM! Thanks @swallez!

**Review by dakrone — APPROVED** [review]
LGTM also, thanks for doing this @swallez!

**Comment by swallez** [comment]
Thanks you both for your patience! Now waiting for CI to hopefully be happy and I'll hit that merge button 😉

**Pull request merged**
